geopandas/sindex.py

```diff
@@ -764,19 +750,40 @@ def query_bulk(self, geometry, predicate=None, sort=False):
             return res
 
         @doc(BaseSpatialIndex.nearest)
-        def nearest(self, geometry):
+        def nearest(
+            self, geometry, return_all=True, max_distance=None, return_distance=False
+        ):
             if not compat.PYGEOS_GE_010:
                 raise NotImplementedError("sindex.nearest requires pygeos >= 0.10")
 
             geometry = self._as_geometry_array(geometry)
-            return super().nearest(geometry)
 
-        @doc(BaseSpatialIndex.nearest_all)
-        def nearest_all(self, geometry, max_distance=None, return_distance=False):
-            geometry = self._as_geometry_array(geometry)
-            return super().nearest_all(
+            if not return_all and max_distance is None and not return_distance:
+                return super().nearest(geometry)
+
+            result = super().nearest_all(
                 geometry, max_distance=max_distance, return_distance=return_distance
             )
+            if return_distance:
+                indices, distances = result
+            else:
+                indices = result
+
+            if not return_all:
+                # first subarray of geometry indices is sorted, so we can use this
+                # trick to get
+                uniques = np.diff(indices[0, :])
+                # always select the first element
+                uniques = np.insert(uniques, 0, 1)
+
+                indices = indices[:, uniques]
```

Comment:
Provided I follow, this actually has a distinct advantage over `pygeos.nearest` in the case of equidistant results in the tree: this is deterministic.  It returns the lowest index out of each set of results per input geometry. (meaning: tests should not fail because of varying result order like we saw in pygeos)

Reply:
Isn't it only deterministic if the output of `nearest_all` is? It doesn't seem the *tree* indices are sorted, so it won't necessarily use the lowest index:

```
In [30]: points = pygeos.points([[1, 1], [0, 1], [1, 0], [0, 0]])

In [31]: tree = pygeos.STRtree(points)

In [32]: tree.nearest_all(pygeos.points([0.5, 0.5]))
Out[32]: 
array([[0, 0, 0, 0],
       [3, 2, 1, 0]])

In [33]: points = pygeos.points([[0, 0], [0, 1], [1, 0], [1, 1]])

In [34]: tree = pygeos.STRtree(points)

In [35]: tree.nearest_all(pygeos.points([0.5, 0.5]))
Out[35]: 
array([[0, 0, 0, 0],
       [0, 2, 1, 3]])
```

It might still be deterministic, depending on how pygeos works (eg in the above, I switched the first and last point in the tree, and it keeps the same order in the output), but so it's not necessarily the lowest index. We could of course first sort if we want to make it deterministically the lowest index.

Reply:
Ah, yes, and I should have known better!  Within pygeos, we build up the tree indices in tree nearest-neighbor traversal order, which is still nondeterministic between GEOS versions (same issue as `nearest`); we don't do a secondary sort on indices per input geometry in the case of multiple equidistant results.  We theoretically could, but I'm not sure we should.